kml/serdes_gpx.cpp

```diff
@@ -0,0 +1,252 @@
+      
+      // Set default pin.
+      if (m_predefinedColor == PredefinedColor::None)
+        m_predefinedColor = PredefinedColor::Red;
+      
+      return true;
+    }
+    return false;
+  }
+  else if (GEOMETRY_TYPE_LINE == m_geometryType)
+  {
+    return m_geometry.IsValid();
+  }
+  
+  return false;
+}
+
+bool GpxParser::Push(std::string_view const & tag)
+{
+  m_tags.push_back(tag);
+  if (GetTagFromEnd(0) == gpx::kWpt)
+    m_geometryType = GEOMETRY_TYPE_POINT;
+  else if (GetTagFromEnd(0) == gpx::kTrkPt || GetTagFromEnd(0) == gpx::kRtePt)
+    m_geometryType = GEOMETRY_TYPE_LINE;
+  return true;
+}
+
+void GpxParser::AddAttr(std::string const & attr, std::string const & value)
+{
+  std::string attrInLowerCase = attr;
+  strings::AsciiToLower(attrInLowerCase);
+  
+  if (GetTagFromEnd(0) == gpx::kWpt)
+  {
+    if (attr == "lat")
+      m_lat = stod(value);
+    else if (attr == "lon")
+      m_lon = stod(value);
+  }
+  else if ((GetTagFromEnd(0) == gpx::kTrkPt && GetTagFromEnd(1) == gpx::kTrkSeg) ||
+           (GetTagFromEnd(0) == gpx::kRtePt && GetTagFromEnd(1) == gpx::kRte))
+  {
+    if (attr == "lat")
+      m_lat = stod(value);
+    else if (attr == "lon")
+      m_lon = stod(value);
+  }
+  
+}
+
+std::string_view const & GpxParser::GetTagFromEnd(size_t n) const
+{
+  ASSERT_LESS(n, m_tags.size(), ());
+  return m_tags[m_tags.size() - n - 1];
+}
+
+void GpxParser::ParseColor(std::string const & value)
+{
+  auto const colorBytes = FromHex(value);
+  if (colorBytes.size() != 3)
+  {
+    LOG(LWARNING, ("Invalid color value", value));
+    return;
+  }
+  m_color = kml::ToRGBA(colorBytes[0], colorBytes[1], colorBytes[2], (char)255);
+}
+
+void GpxParser::Pop(std::string_view const & tag)
+{
+  ASSERT_EQUAL(m_tags.back(), tag, ());
+  
+  if (tag == gpx::kTrkPt || tag == gpx::kRtePt)
+  {
+    m2::Point p = mercator::FromLatLon(m_lat, m_lon);
+    if (m_line.empty() || !AlmostEqualAbs(m_line.back().GetPoint(), p, kMwmPointAccuracy))
+      m_line.emplace_back(p);
+  }
+  else if (tag == gpx::kTrkSeg || tag == gpx::kRte)
+  {
+    m_geometry.m_lines.push_back(std::move(m_line));
+  }
+  else if (tag == gpx::kWpt)
+  {
+    m_org = mercator::FromLatLon(m_lat, m_lon);
+  }
+  
+  if (tag == gpx::kRte || tag == gpx::kTrkSeg || tag == gpx::kWpt)
+  {
+    if (MakeValid())
+    {
+      if (GEOMETRY_TYPE_POINT == m_geometryType)
+      {
+        BookmarkData data;
+        data.m_name = std::move(m_name);
+        data.m_description = std::move(m_description);
+        data.m_color.m_predefinedColor = m_predefinedColor;
+        data.m_color.m_rgba = m_color;
+        data.m_point = m_org;
+        data.m_customName = std::move(m_customName);
+        // Here we set custom name from 'name' field for KML-files exported from 3rd-party services.
+        if (data.m_name.size() == 1 && data.m_name.begin()->first == kDefaultLangCode &&
+            data.m_customName.empty())
```

Comment:
Does it fit in 120 characters?
```suggestion
        if (data.m_name.size() == 1 && data.m_name.begin()->first == kDefaultLangCode && data.m_customName.empty())
```

Reply:
Yes, fixed